ft/2124 Secure uploaded contracts

martynasracct opened the pull request:
For reviewers:
  - 8402d67676956b9c0905b218a2dad54b97df3f8c - adds missing tests for `can_view_agreement`.
  - 83109e815aad6900598f78b128c42c094dc8b2c3 - introduce new logic for uploading private files: new settings, new storage, etc.
  - e20736d29f238b874917ee2c6132625a21f378ca - adds `AgreementFileDownloadView`.
  - e7d67de9b4fe16ae9de0985fcab83bcd3f12c82c - fix test file uploading to `INTERNAL_MEDIA_ROOT`.
  - 156a1eaa3737940c55d1b64193727e00d060c725 - switch to using `@override_settings(DEBUG=...)` in tests.
  - 2f636fe902f286987eaf559a9a9038633718c2e0, 973978e5f68b71f55b30286ae7b08121ba3d452b, 82d0a710c80896441790e366d989aaefcaf1cdaa, f5b35be9e968070cd415c647692263ea2fdca5a4, 4bfeb6bb174704012d5bb17e523bc62b97d5b233, d76b0d7a4e70b0f1353438c1af282495e9e166e9, 66abbc10addc04d03304d8a01d762c3b7d837b2f, dbdb4ea763bdd2800e373bbf5f98eaf10479e325 - fixes for missing template variables that started failing random tests after introducing `@override_settings(DEBUG=...)`.
  - 7a7929ee3bdaaf561f7a2ae3aa95b9848b575868 - small PR fixes, added CHANGES.rst

**Summary:**
* Add a way to upload files to another directory, that should not be publicly available, using `INTERNAL_MEDIA_URL`, `INTERNAL_MEDIA_ROOT` and `internal_media_storage`. Changed `SmartContractFile` uploads to this directory.
* Add endpoint `AgreementFileDownloadView` to download uploaded contract files:
  * If `DEBUG=True` - file is downloaded same way as before.
  * If `DEBUG=False` - file path is returned via `X-Accel-Redirect`. This header tells Nginx to do redirect to file and hides file path.
* `AgreementFileDownloadView` has `can_view_agreement` permissions.

**Note:**
Nginx must be configured to serve  `INTERNAL_MEDIA_URL` files internally:
```
    location /internal-media {
        internal;
        alias local/path/to/internal-media/dir;
    }
```

**Ticket:**
https://github.com/atviriduomenys/katalogas/issues/2124

martynasracct committed:
2124 Add tests for can_view_agreement

martynasracct committed:
2124 Introduce INTERNAL_MEDIA_ROOT. Upload contract files to this dir

Works same way as MEDIA_ROOT. Separate dir is needed so that we can tell Nginx that files in this dir must only be accessible internally

martynasracct committed:
2124 Add AgreementFileDownloadView to download agreement

gts446 commented on vitrina/smart_contracts/views.py:
```diff
@@ -548,3 +550,26 @@ def _perform_action(self, form: ModelForm) -> None:
+
+        if settings.DEBUG:
+            return FileResponse(open(agreement_file.file.path, "rb"), as_attachment=True)
+        else:
```
No need for `else` here

martynasracct replied:
Fixed in 7a7929ee3bdaaf561f7a2ae3aa95b9848b575868

Struchid reviewed (approved):
Clean, good work!:P

gts446 commented on tests/smart_contracts/test_views.py:
```diff
@@ -1995,3 +1996,171 @@ def test_multiple_pdf_files_found(
+    def test_return_file_as_attachment_if_debug_true(
+        self, app: DjangoTestApp, agreement_pdf: Path
+    ):
+        debug = settings.DEBUG
```
Probably nicer and safer option would be to use `@override_settings(DEBUG=True)` decorator. Because in current case, if test fails `settings.DEBUG` is not reset to initial value.

martynasracct replied:
I agree with you, but when I use it - bunch of unrelated tests breaks, because of some custom template tag, that is (once again) completely unrelated to my PR 

https://github.com/atviriduomenys/katalogas/actions/runs/19853743825

I am looking into possible ways to fix it now

martynasracct replied:
Fixed in 156a1eaa3737940c55d1b64193727e00d060c725

martynasracct committed:
Fix for test uploaded files in INTERNAL_MEDIA_ROOT

Files created by tests were beign creted in INTERNAL_MEDIA_ROOT directory, because INTERNAL_MEDIA_ROOT and INTERNAL_MEDIA_URL values were set on model load, and test fixtures had no effect on it. When variables set inside class methods - they are not initialized together with storage itself and can be changed via test fixtures

martynasracct committed:
Use @override_settings(DEBUG=...) in tests and fix breaking tests

When running whole test suite after adding 2 tests with @override_settings(DEBUG=...), ~100 random tests started breaking. Error:
FAILED tests/structure/test_views.py::test_property_enum_item_update - Failed: Undefined template variable 'logged_in_user' in '/home/runner/work/katalogas/katalogas/vitrina/templates/component/navbar.html'

Turns out, these tests break if running with DEBUG=True, because logged_in_user variable is used in navbar.html which is one of base HTMLs, but it is set in only one view. Turns out it's not neede at all.

martynasracct committed:
Removed langage change form parameter that has no value

{{ redirect_to }} template variable is never set. This parameter makes a lot if tests to fail with error:
ailed: Undefined template variable 'redirect_to' in '/home/runner/work/katalogas/katalogas/vitrina/templates/component/navbar.html'

martynasracct committed:
Disable CMS toolbar for all tests.

martynasracct committed:
Remove template variable {{ total_count }}. Value is set using Ajax call

martynasracct committed:
Remove unused template variable {{ organization.title }}. Use {{ dataset.organization.title }} instead

martynasracct committed:
Add missing {{ current_title }} template variable. Also add translations

martynasracct committed:
Fix missing {{ prop.metadata.first.type }} template variable. Don't call it if not set.

martynasracct committed:
Remove unused template varialbe {{ rating.description }}

martynasracct committed:
Remove unused template variable {{ dataset.distributionConditions }}. This model field is removed.

martynasracct committed:
Small PR fixes. Add CHANGES.rst

Struchid commented on vitrina/storages.py:
```diff
@@ -0,0 +1,15 @@
+        return settings.INTERNAL_MEDIA_ROOT
+
+    @property
+    def base_url(self):
```
Typehints! :p

martynasracct replied:
🦅👁️! Fixed in 42ad88ccadb97db92804d90e523760a33f614a1d

Struchid commented on vitrina/locale/en/LC_MESSAGES/django.po:
```diff
@@ -4958,12 +4967,33 @@ msgstr "Answer to"
+msgstr "Delete comment?"
+
+msgid "Šis veiksmas negrąžinamas."
+msgstr ""
```
Missing translation ;/

martynasracct replied:
Fixed in 42ad88ccadb97db92804d90e523760a33f614a1d

Struchid commented on tests/smart_contracts/test_views.py:
```diff
@@ -1995,3 +1995,165 @@ def test_multiple_pdf_files_found(
+            reverse("agreement-file-download", args=[agreement.project.pk, agreement.pk, agreement_file.pk])
+        )
+
+        assert response.status_code == 200
```
Could we add an additional check here? I've encountered that sometimes you still get a http 200, while getting a form error of some kind, not sure if it's the same when you call via API or just execute the form code.

martynasracct replied:
This few has no form, only get without much logic. There are other tests that checks if response is correct based on DEBUG settings. I think asserting only status code here is sufficient enough

martynasracct committed:
Small PR fixes

Gustascct commented on vitrina/structure/templates/vitrina/structure/model_data_table.html:
```diff
@@ -106,7 +106,7 @@ <h2 class="title is-4-mobile">{% translate "Filtras" %}:</h2>
     </div>
 
     <div class="mt-5 mb-2">
-        <p id="total_count_id">{{ total_count }}</p>
+        <p id="total_count_id"></p>
     </div>
```
Does this do anything after you removed `{{ total_count }}`? If its just an empty paragraph maybe remove it entirely?

martynasracct replied:
Yes. There is ajax request in different HTML that sets value to this element

https://github.com/atviriduomenys/katalogas/blob/c1d4865f49758dff05ef6553049ad79469889a2f/vitrina/structure/templates/vitrina/structure/model_data.html#L75-L87

Gustascct replied:
Okay nice
